File: 17. 电话号码的字母组合/swift/main/main/main.swift
//
//  main.swift
//  main
//
//  Created by cedricwu on 6/2/25.
//

import Foundation

class Solution {
    let map = [
        "0": "",
        "1": "",
        "2": "abc",
        "3": "def",
        "4": "ghi",
        "5": "jkl",
        "6": "mno",
        "7": "pqrs",
        "8": "tuv",
        "9": "wxyz"
    ]
    var result = [String]()
    var path = [String]()
    
    func letterCombinations(_ digits: String) -> [String] {
        guard digits.count > 0 else {
            return []
        }
        
        backtracing(digits: Array(digits), index: 0)
        
        return result
    }
    
    private func backtracing(digits: [String.Element], index: Int) {
        if digits.count == path.count {
            let item = path.joined(separator: "")
            result.append(item)
            return
        }
        let digit = String(digits[index])
        let list = Array(map[digit]!)
        for i in 0 ..< list.count {
            let item = String(list[i])
            path.append(item)
            backtracing(digits: digits, index: index + 1)
            _ = path.popLast()
        }
    }
}

let s = Solution()
print(s.letterCombinations("23"))

File: 17. 电话号码的字母组合/rust/main/src/main.rs
struct Solution {}

const map: [&str; 10] = [
    "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz",
];

impl Solution {
    fn back_trace(result: &mut Vec<String>, s: &mut String, digits: &String, index: usize) {
        let len = digits.len();
        if len == index {
            result.push(s.to_string());
            return;
        }
        let digit = (digits.as_bytes()[index] - b'0') as usize;
        for i in map[digit].chars() {
            s.push(i);
            Self::back_trace(result, s, digits, index + 1);
            s.pop();
        }
    }

    pub fn letter_combinations(digits: String) -> Vec<String> {
        if digits.is_empty() {
            return vec![];
        }
        let mut res = vec![];
        let mut s = String::new();
        Self::back_trace(&mut res, &mut s, &digits, 0);
        res
    }
}

fn main() {
    println!("Hello, world!");
}
